pkg/compiler/check.go

```diff
@@ -614,7 +632,12 @@ func (comp *compiler) checkTypeCtors(t *ast.Type, dir prog.Dir, isArg bool,
 		}
 		checked[key] = true
 		for _, fld := range s.Fields {
-			comp.checkTypeCtors(fld.Type, dir, false, ctors, inputs, checked)
+			// If the field has direction, it overrides any upper levels.
+			if fldDir, fldHasDir := comp.genFieldDir(fld); fldHasDir {
```

Comment:
Will be less logic duplication with:
```
fldDir, fldHasDir := comp.genFieldDir(fld)
if !fldHasDir {
    fldDir = dir
}
comp.checkTypeCtors(fld.Type, fldDir, false, ctors, inputs, checked)
```

Reply:
This is right, [done](https://github.com/google/syzkaller/pull/2008/files#diff-47a4f07e1880cd23bc05dca73d329c1dR635-R639).